DC-1259: Enable listFiles to work for Azure; Add try/catch to avoid issue in future

snf2ye opened the pull request:
__Jira ticket__: https://broadworkbench.atlassian.net/browse/DC-1259

## Addresses
Users have reported issues with the listFiles endpoint on Azure. This PR addresses the root issue (commit 1) and a "fix" that enable the listFiles endpoint to work for datasets that have already encountered this issue (commit 2). 

The reported issue: listFiles does not reliably work. 
User reported that they saw errors when they used the offset/limit options to retrieve files beyond what exists. 

For example - If there are 25 files:
✅  list files, offset: 0, limit:25
❌  list files, offset: 0, limit:26
- Fails with the following error:` 404 The specified resource does not exist`
- We found logs with: `Error retrieving file metadata for file: <nonexistent fileId>`
- This corresponded to an error from:
         - `TableFileDao.retrieveFileMetadata()`
         - Called by `TableFileDao.batchRetrieveFileMetadata`
         - Which is called based on a list of directory entries retrieved by `directoryDao.enumerateFileRefEntries`.

Conclusion: This dataset erroneously has directory entires listing fileIds that do not exist in the dataset. 

How did this happen? The user reported that there were failed file ingests on this dataset. So, possibly these are fileIds left over from failed ingests.

**Failure on Ingest**

Took some tracking down, but found one of the failed FileIngestWorkerFlights for the user’s case - Flight Id: nzrj0brYRDKgrM5y2XG6jQ (the logs reference datarepo34f9c0d53a784e7d85b52089280ff87adataset, which includes the dataset id, and other clues such as the file names).

“DO” step of IngestFileAzurePrimaryDataStep failed with Interrupted exception on attempting to copy file 
<img width="1072" alt="image-20240916-183857" src="https://github.com/user-attachments/assets/aecccfd5-f0ff-46fe-b502-d73d707cc533">


“UNDO” step of IngestFileAzurePrimaryDataStep dismally failed when attempting to delete the file 
<img width="914" alt="image-20240916-184102" src="https://github.com/user-attachments/assets/331a181e-b4ac-4cf9-b8d4-45092a506a16">

Dismal Failure:
<img width="964" alt="image-20240916-184130" src="https://github.com/user-attachments/assets/5dd0bda5-6139-43dd-9ced-075d53253335">


## Summary of changes
- Make "undo" step of IngestFileAzurePrimaryDataStep try/catch on deleting the file, so that we do not dismally fail on undo. This should enable the cleanup to remove the other file metadata entries
- Ignore case where directory entries do not have matching file entry so that listFiles endpoint should still work even if the dataset has run into this error case. 

## Testing Strategy
- Unit tests
- Hard to reproduce exact issue - we should see whether it fixes the issue for our users

snf2ye committed:
catch exceptions on undo so that file metadata is cleaned up rather than dismally failing

spotless

snf2ye committed:
Handle mismatched directory file and file lists - this will enable listFiles to work for existing azure datasets

- In order to avoid making changes for GCP, add enforceMatchingDirectoryEntries flag so that these changes only apply to Azure
- Update unit tests

s-rubenstein commented on src/main/java/bio/terra/service/filedata/flight/ingest/IngestFileAzurePrimaryDataStep.java:
```diff
@@ -100,8 +100,9 @@ public StepResult undoStep(FlightContext context) {
-        fileId, fileName, storageAccountResource, userRequest)) {
+    try {
+      azureBlobStorePdao.deleteDataFileById(fileId, fileName, storageAccountResource, userRequest);
+    } catch (Exception ex) {
```
Do you know a more specific exception we could use here? I am always a little worried when it is just `catch(Exception`.

snf2ye replied:
The specific exception we ran into in the error was an Interrupted exception. But, I want the undo to still pretty broadly pass no matter the exception so that the rest of the cleanup in the flight still occurs. I think we should keep this broad catch exception.

rjohanek commented on src/main/java/bio/terra/service/filedata/FileMetadataUtils.java:
```diff
@@ -196,17 +195,23 @@ public static Md5ValidationResult validateFileMd5ForIngest(
+      String collectionId,
+      boolean enforceMatchingDirectoryEntries) {
+    if (enforceMatchingDirectoryEntries && (directoryEntries.size() != files.size())) {
       throw new FileSystemExecutionException("List sizes should be identical");
```
could this exception description be more helpful to a user? like they may not know what the referenced lists are

rjohanek commented on src/main/java/bio/terra/service/filedata/FileMetadataUtils.java:
```diff
@@ -196,17 +195,23 @@ public static Md5ValidationResult validateFileMd5ForIngest(
+                      .orElseThrow(
+                          () ->
+                              new FileSystemExecutionException(
+                                  "There should be a matching directory entry for every file."));
```
why add this exception? and do we only want to throw it if the enforceMatchingDirectoryEntries flag is true?

snf2ye replied:
No, the enforceMatchingDirectoryEntries only applies to directory entries that are missing the underlying file metadata entries. 
Why?
- We use the directory entry to look up the file entry. If this file entry exists, then there must be a matching directory entry. 
<img width="1039" alt="image" src="https://github.com/user-attachments/assets/fe5d4303-489d-46d5-8a87-089765b42580">

However, (and this is the case we're accounting for) there could exist a directory entry and then we go to look up the file, no file is found.

rjohanek commented on src/main/java/bio/terra/service/filedata/azure/tables/TableDao.java:
```diff
@@ -434,6 +437,9 @@ private FSItem makeFSFile(
     FireStoreFile fireStoreFile =
         fileDao.retrieveFileMetadata(datasetTableServiceClient, datasetId, fileId);
+    if (fireStoreFile == null) {
+      throw new FileSystemExecutionException("Error retrieving file metadata for file");
```
is this a separate issue? missing metadata compared to directory entries or is this the same?

snf2ye replied:
Yeah, exactly: Missing the file metadata (vs. directory entries).
This is just handling the case that the file metadata isn't returned for the given fileId. (which we're now allowing rather than throwing an exception)

rjohanek reviewed (approved):
overall, looks good and what a great PR description thank you! super helpful. I just have a few questions.

snf2ye committed:
Improve Performance of toFileModel() call

@pshapiro4broad brought up a good point that the proposed change would not be as performant of the original (O(n) vs. O(n^2)). By first creating a map between the file id and the directory entry, we can make some improvements here.

At the end of the day, I don't expect for this code to be under heavy load: It is only used by the listFiles endpoint which only returns paged results.
However, this was a minor fix to help get the performance back in the same ballpark!

snf2ye committed:
Updated error messages

s-rubenstein reviewed (approved):
LGTM

snf2ye committed:
raise level of logging for test run to warn so that the link to the gradle scan isn't truncated

fboulnois commented on .github/workflows/int-and-connected-test-run.yml:
```diff
@@ -74,7 +74,7 @@ jobs:
           base64 --decode <<< ${{ secrets.SA_B64_CREDENTIALS }} > ${GOOGLE_APPLICATION_CREDENTIALS}
           base64 --decode <<< ${{ secrets.B64_RBS_APPLICATION_CREDENTIALS }} > ${RBS_CLIENT_CREDENTIAL_FILE_PATH}
           # assemble code and run connected tests
-          ./gradlew --scan assemble testConnected
```
Is the order important? I can't remember. Also Phil mentioned we could probably drop the `assemble` because `testConnected` builds the code.

snf2ye replied:
I don't think that order matters.  I moved scan to the end mostly because it matches what we have in the datarepo-actions repo and I didn't want to have to iterate on it given the long test run time, but it looks like it matches what is in the [gradle docs](https://docs.gradle.org/current/userguide/build_scans.html#enable_build_scans). 
And, I think Phil is right, we can remove assemble. I'm going to leave so that I don't have to wait for another test run, but I'll remove it with my connected test work.

fboulnois reviewed (approved):
Looks reasonable, one minor question in relation to the Action changes:
